### DISCUSSION TITLE
Problem with pad in transforms

### DISCUSSION THREAD

**Discussion by XIANYUNYEHE-DEL** [Project-MONAI/MONAI#4680] [Q&A] [answered]
Hi,
I have an another kind of label need to be pad. I tried three different pad methods offered by the authorities.But the final result was problematic and random. For example, I use this code to pad.
```
transforms.DivisiblePadd(keys=["image", "label", "std"], k=96),
transforms.RandSpatialCropd(keys=["image", "label", "std"],
                                        roi_size=[96,96,96], 
                                        random_size=False),
```
Sometimes it happens
```
> collate dict key "image" out of 9 keys
>> collate/stack a list of tensors
> collate dict key "label" out of 9 keys
>> collate/stack a list of tensors
> collate dict key "std" out of 9 keys
>> collate/stack a list of tensors
>> E: stack expects each tensor to be equal size, but got [5, 96, 96, 96] at entry 0 and [5, 96, 96, 31] at entry 1, shape [torch.Size([5, 96, 96, 96]), torch.Size([5, 96, 96, 31])] in collate([tensor([[[[ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
          [ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
          [ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
          ...,
          [ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
          [ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
          [ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000]],
.....
```
Images and labels are one channel. Std is an another kind of label which has the same size with image but has 5 channels. The problem only occurs in std.
I don't know how to solve this problem. I am sure that there is no problem with the format of the data. Is there anything wrong with my usage?

**Comment by wyli** [comment, marked as answer]
the code example looks fine, according to the log, the output 'std' is sometimes `torch.Size([5, 96, 96, 96])` and sometimes `torch.Size([5, 96, 96, 31])`. if you could provide more info, we can discuss it further here.

**Reply by XIANYUNYEHE-DEL** [reply]
Thank you for your reply. The problem is that I want to keep the minimum size of 'std' at `torch.Size([5, 96, 96, 96])`. But it looks like `transforms.DivisiblePadd` doesn't work on 'std'. After using `transforms.RandSpatialCropd`, some 'std' size still less than `torch.Size([5, 96, 96, 96])` like `torch.Size([5, 96, 96, 31])`. So, I want to make 'std' which minimum size less than `torch.Size([5, 96, 96, 96])` have the same pad-operation as the image and label.

**Reply by wyli** [reply]
ok, does std/image/label always have the same initial shape? if the shapes are the same, they will be transformed consistently, your example should be fine. if the shapes are different, you should preprocess them to have the same shape before these transforms.

**Reply by XIANYUNYEHE-DEL** [reply]
Thanks for your advice. After inspection, there is a problem of different shapes in my data. I think that led to my mistake.

**Reply by wyli** [reply]
sure, you might want to try `Resized` or `ResizeWithPadOrCropd` to make the shapes consistent.

in fact it also works fine if everything is smaller than 96 (I put a testing script below), but I don't think it makes much sense to use divisiblepad if that's the case, perhaps `ResizeWithPadOrCropd` is static and more robust.

```py
import numpy as np

from monai.transforms import Compose, DivisiblePadd, RandSpatialCropd

t = Compose(
    [
        DivisiblePadd(keys=["image", "label", "std"], k=96),
        RandSpatialCropd(keys=["image", "label", "std"], roi_size=[96, 96, 96], random_size=False),
    ]
)

inputs = {"image": np.zeros((1, 95, 94, 95)), "label": np.zeros((1, 93, 94, 95)), "std": np.zeros((1, 95, 96, 31))}

for _ in range(10):
    out = t(inputs)
    print("image", out["image"].shape, "label", out["label"].shape, "std", out["std"].shape)
```

```
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
image (1, 96, 96, 96) label (1, 96, 96, 96) std (1, 96, 96, 96)
```

**Reply by XIANYUNYEHE-DEL** [reply]
I think you are right. I will try to use `ResizeWithPadOrCropd` method. In the meantime, thanks for your help and your nice work of Monai.
